### DISCUSSION TITLE
How to Measure the SSR Response Time

### DISCUSSION THREAD

**Discussion by emrekardaslar** [remix-run/remix#4098] [Q&A]
Hello everyone. I want to measure the SSR response time, I could measure the time passed for hydration by just using two Date objects before and after hydrate method in entry.client.tsx. How can I measure the time passed in SSR inside the application? I tried the same way by starting the time at the start of the Document method and print it at the end but it seems wrong.

`function Document({children,title,}: {children: React.ReactNode; title?: string;}) {
    const start = Date.now();
    ....
    console.log('Total time', Date.now() - start)`

**Comment by emrekardaslar** [comment]
I've tried another way by adding new code to the entry.server.tsx:

```import { renderToString } from "react-dom/server";
import { RemixServer } from "remix";
import type { EntryContext } from "remix";

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const start = Date.now();
  let markup = renderToString(
    <RemixServer context={remixContext} url={request.url} />
  );

  responseHeaders.set("Content-Type", "text/html");
  responseHeaders.set("Set-Cookie", `response-time=${(Date.now()-start).toString()}`)

  return new Response("<!DOCTYPE html>" + markup, {
    status: responseStatusCode,
    headers: responseHeaders
  });
}
```

Do you think it is a good way to solve to measure total time spent in server side? Is there another way?

**Reply by kiliman** [reply]
Not really. By the time `handleRequest` is called in *entry.server*, Remix has already processed all the loaders/actions. This function is simply to generate the final response.

The correct way to get the starting time for the request is to add `getLoadContext` in ~~*remix.config*~~ *server.js*. That's called just before Remix starts processing the request.

EDIT: Corrected filename

https://remix.run/docs/en/v1/api/conventions#loader-context

**Reply by emrekardaslar** [reply]
Hello @kiliman , first of all, thanks for answering my question. I couldn't understand where should I exactly put getLoadContext in remix.config.js. Is there any example for that?

Thanks!

**Reply by emrekardaslar** [reply]
I saw an example which is made by you (https://codesandbox.io/s/f8s1s0?file=/server/index.js:291-293) I see that this example uses express server. I'm not using express server or anything else (using remix app server). How can I make the measurement? Should I measure the time passed inside the loader?

Thanks!

**Reply by kiliman** [reply]
Sorry, that was my mistake. `getLoadContext` is actually in your server entry file. You would need to use the Express adapter. If you're using Remix App Server then I wrote a script to eject from RAS to Express. https://rmx.fyi/ejectras

**Reply by emrekardaslar** [reply]
Is there a way to measure it without using express or etc. just by using Remix App Server?

**Reply by kiliman** [reply]
Not really, since RAS doesn't expose it's configuration. You would have to patch it to add your function. Take a look at `Metronome` which is a performance monitoring solution for Remix. You can see how they set it up. https://metronome.sh/docs/getting-started/installing-metronome

**Reply by sergiodxa** [reply]
> Is there a way to measure it without using express or etc. just by using Remix App Server?

You could measure each loader time individually, by doing

```ts
export async function loader(args: LoaderArgs) {
  let start = Date.now()
  try {
    // code here
  } finally {
    let end = Date.now()
    console.log(`Total time ${end - start}ms`)
  }
}

**Reply by emrekardaslar** [reply]
Thanks for the reply! I will use that way for measurement for now.
